# Análise de Dados - E-commerce Olist

## Exploração inicial dos dados

In [1]:
import pandas as pd

## Importação dos dados

In [2]:
df_clientes = pd.read_csv("../data/raw/olist_customers_dataset.csv")

df_pedidos = pd.read_csv("../data/raw/olist_orders_dataset.csv")

df_itens = pd.read_csv("../data/raw/olist_order_items_dataset.csv")

df_pagamentos = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")

df_avaliacoes = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")

df_produtos = pd.read_csv("../data/raw/olist_products_dataset.csv")

df_categorias = pd.read_csv("../data/raw/product_category_name_translation.csv")

## Visão geral dos datasets

In [3]:
df_clientes.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [4]:
df_pedidos.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [5]:
df_produtos.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


## Informações gerais dos datasets

In [6]:
tabelas = {
    "Clientes": df_clientes,
    "Pedidos": df_pedidos,
    "Itens": df_itens,
    "Pagamentos": df_pagamentos,
    "Avaliações": df_avaliacoes,
    "Produtos": df_produtos,
    "Categorias": df_categorias
}

for nome, df in tabelas.items():
    print(f"{nome}")
    print(f"Linhas: {df.shape[0]}")
    print(f"Colunas: {df.shape[1]}")
    print("-" * 15)

Clientes
Linhas: 99441
Colunas: 5
---------------
Pedidos
Linhas: 99441
Colunas: 8
---------------
Itens
Linhas: 112650
Colunas: 7
---------------
Pagamentos
Linhas: 103886
Colunas: 5
---------------
Avaliações
Linhas: 99224
Colunas: 7
---------------
Produtos
Linhas: 32951
Colunas: 9
---------------
Categorias
Linhas: 71
Colunas: 2
---------------


## Estrutura dos datasets

In [7]:
df_clientes.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


In [8]:
df_pedidos.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


In [9]:
df_itens.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 6.0 MB


In [10]:
df_pagamentos.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


In [11]:
df_avaliacoes.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 5.3 MB


In [12]:
df_produtos.info()

<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB


In [13]:
df_categorias.info()

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 1.2 KB


## Qualidade dos dados

Nesta etapa, vamos verificar a presença de valores ausentes, registros duplicados e possíveis inconsistências nos datasets antes de iniciar o tratamento dos dados.

### Valores ausentes

In [14]:
for nome, df in tabelas.items():
    print(f"\n{nome}")
    print(df.isnull().sum())


Clientes
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Pedidos
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Itens
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Pagamentos
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Avaliações
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_

### Análise de valores ausentes

A análise de valores ausentes identificou problemas em três datasets.

No `df_pedidos`, foram encontrados valores ausentes nas colunas `order_approved_at` (160), `order_delivered_carrier_date` (1.783) e `order_delivered_customer_date` (2.965). Essas colunas representam diferentes etapas do processamento e da entrega dos pedidos, portanto os valores ausentes deverão ser investigados considerando o `order_status` antes de definir uma estratégia de tratamento.

No `df_avaliacoes`, foram identificados 87.656 valores ausentes em `review_comment_title` e 58.247 em `review_comment_message`. Como essas colunas representam comentários opcionais dos clientes, os valores ausentes não serão tratados automaticamente como erros.

No `df_produtos`, foram encontrados 610 valores ausentes nas informações de categoria e nas características textuais dos produtos, além de 2 valores ausentes nas dimensões e no peso dos produtos. Esses registros deverão ser avaliados durante a etapa de tratamento.

Os demais datasets (`df_clientes`, `df_itens`, `df_pagamentos` e `df_categorias`) não apresentam valores ausentes.

### Análise de registros duplicados

Nesta etapa, será verificada a existência de registros duplicados nos datasets, a fim de identificar possíveis repetições que possam afetar as análises posteriores.

In [15]:
for nome, df in tabelas.items():
    print(f"{nome}: {df.duplicated().sum()} registros duplicados")

Clientes: 0 registros duplicados
Pedidos: 0 registros duplicados
Itens: 0 registros duplicados
Pagamentos: 0 registros duplicados
Avaliações: 0 registros duplicados
Produtos: 0 registros duplicados
Categorias: 0 registros duplicados


### Resultado da análise de registros duplicados

A análise não identificou registros completamente duplicados em nenhum dos datasets analisados. Portanto, não foram encontrados registros que precisem ser removidos por duplicidade nesta etapa.

É importante destacar que a ausência de linhas completamente duplicadas não significa que todos os identificadores sejam únicos em cada tabela. A repetição de identificadores pode ser esperada devido aos relacionamentos entre os datasets e será analisada posteriormente.

### Análise dos tipos de dados

Nesta etapa, serão avaliados os tipos de dados das colunas para identificar campos que possam exigir conversão ou padronização antes das análises.

In [16]:
for nome, df in tabelas.items():
    print(f"\n{nome}")
    print(df.dtypes)


Clientes
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

Pedidos
order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

Itens
order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

Pagamentos
order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object

Avaliações
review_id                    str
order_id          

### Resultado da análise dos tipos de dados

A análise dos tipos de dados identificou que as principais colunas que exigem atenção são as relacionadas a datas.

No `df_pedidos`, as colunas `order_purchase_timestamp`, `order_approved_at`, `order_delivered_carrier_date`, `order_delivered_customer_date` e `order_estimated_delivery_date` estão armazenadas como `str`, embora representem datas e horários.

No `df_itens`, a coluna `shipping_limit_date` também está armazenada como `str`.

No `df_avaliacoes`, as colunas `review_creation_date` e `review_answer_timestamp` estão armazenadas como `str`.

Essas colunas deverão ser convertidas para um tipo de data apropriado durante a etapa de tratamento dos dados, permitindo a realização de análises temporais e cálculos relacionados aos prazos de entrega.

As principais colunas numéricas, como `price`, `freight_value`, `payment_value` e `review_score`, já estão armazenadas em tipos numéricos adequados. Os identificadores dos registros estão armazenados como `str`, o que é apropriado para esse tipo de informação.

## Conclusão da exploração inicial

A exploração inicial permitiu compreender a estrutura dos sete datasets utilizados no projeto e identificar os principais pontos que deverão ser considerados nas etapas seguintes.

Foram analisadas as dimensões e estruturas das tabelas, além da presença de valores ausentes, registros duplicados e tipos de dados.

Foram identificados valores ausentes principalmente nos datasets de pedidos, avaliações e produtos. Também foi observado que as colunas relacionadas a datas estão armazenadas como texto e deverão ser convertidas para um formato adequado durante o tratamento dos dados.

Não foram identificados registros completamente duplicados nos datasets analisados.

A exploração inicial também permitiu compreender algumas características dos dados, como os diferentes status dos pedidos.

Com essa etapa concluída, os dados podem seguir para a etapa de tratamento, na qual serão realizadas as correções e padronizações necessárias para preparar as informações para as análises de negócio.